### DISCUSSION TITLE
Constructing a custom problem in Pymoo (Picking a subset of Vehicles from a Bag)

### DISCUSSION THREAD

**Discussion by usmanustc** [anyoptimization/pymoo#397] [Q&A]
I am trying to follow this [method (https://pymoo.org/customization/custom.html) to solve an optimization problem. I have a bag of N=100 vehicles which have certain properties. For example cost, rank, and mpg. I am looking to find k=5 subsets of vehicles that have higher mpg and rank than others. N and k can take any values.

Following the (pymoo)[https://pymoo.org/customization/custom.html], the evaluation function looks like this:

    def _evaluate(self, x, out, *args, **kwargs):
        out["F"] = np.array([x["rank"], x["mpg"]], dtype=float)

Here, x has as many dimensions as many columns for vehicles. In this case, I am using rank and mpg as I want to maximize these.

The guide says to create methods including:

Sampling Crossover Mutation

I want to know how can I formulate the above problem in these functions.

Sampling: refer the link above, should I create a method that takes 5 random vehicles from bag and return these?
Crossover: I am blind to this. What should I be doing in my case?
Mutation: I don't know what to implement here.
If somebody has used pymoo, give me some hints on the above so that I can do my implementation.

Many Thanks

**Comment by cyrilpic** [comment]
You are trying to solve a subset selection problem. That means that you are looking for k indices from N. As such, your decision vector `x` is a list with k integers (ideally you would like to use a set instead of a list but I'm not sure that pymoo can do that), e.g. `x=[10, 32, 55, 97, 98]`. From that, you need to think about methods that can make changes to this list.

Sampling: randomly choose k from N.
Mutation: randomly replace one element with an index that is not yet in the list.
Crossover: given two lists, swap some elements between the lists.

I also want to point out that your evaluate method should be a bit different:

```python
def _evaluate(self, x, out, *args, **kwargs):
    out["F"] = np.array([self.metrics["rank"][x], self.metrics["mpg"][x]], dtype=float)
```
where metrics is the full list from which you want to pick your subset `x`.

**Reply by cyrilpic** [reply]
I think you are missing some sum operations in the _evaluate function. Can you print the shape of `out["F"]`?

**Reply by usmanustc** [reply]
I have removed the previous code and error screenshot to avoid confusion. I am going to add new version now with the results, so that its clear.

**Reply by usmanustc** [reply]
Hi,

Here is my data:

[9,13],[9,14],[7,15],[6,12],[8,16],[8,14],[6,12],[3,18],[4,14],[5,14],[8,11],[8,18]

I have added it to present a simpler case. First is rank, other is mpg. I want to select any 5 options from the above which give me maximum for rank and mpg. For instance, 9 is max for the rank, and 18 is the highest for mpg. I want to get an optimal list of subsets where both are higher. This example however implements MIN. I guess, I can use a -ve sign to make it a MAX case. But for now, I want to resolve the issues I m getting in running this.

I have set n_max=5, because I need a subset of size 5. I pass L to the problem. The Evaluate function is now changed like you said. Whatever comes to it, it takes a sum for rank and mpg, and stores in out["F"]. So if this is passed:

[ 2  3  4 10  1]

out["F"] = [38, 68]

First one is sum of rank of items 2,3,4,10,1 in L. Second is the sum of mpg of these items.

I am not sure if it's correct. Here we are saying that we are minimizing the sum of all subset items. Is that correct?

I am attaching the code in separate file, and I have screenshot of results, and an error that popsup when one iteration is completed.
![WeChat Screenshot_20230322111742](https://user-images.githubusercontent.com/71557427/226793407-d569335e-989f-46a4-b7b5-c027dffaa4fe.png)
![WeChat Screenshot_20230322111809](https://user-images.githubusercontent.com/71557427/226793418-495e08b2-9cc3-4533-b1d6-92350412dc1e.png)
[code-ga.txt](https://github.com/anyoptimization/pymoo/files/11035851/code-ga.txt)

**Reply by blankjul** [reply]
```
import numpy as np

from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.core.crossover import Crossover
from pymoo.core.mutation import Mutation
from pymoo.core.problem import ElementwiseProblem
from pymoo.core.sampling import Sampling
from pymoo.optimize import minimize


class SubsetProblem(ElementwiseProblem):
    def __init__(self, L, n_max):
        super().__init__(n_var=len(L), n_obj=2, n_ieq_constr=0)
        self.L = L
        self.n_max = n_max

    def _evaluate(self, x, out, *args, **kwargs):
        out["F"] = self.L[x].sum(axis=0).astype(float)


class MySampling(Sampling):

    def _do(self, problem, n_samples, **kwargs):
        X = np.full((n_samples, problem.n_var), False, dtype=bool)
        for k in range(n_samples):
            I = np.random.permutation(problem.n_var)[:problem.n_max]
            X[k, I] = True

        return X


class BinaryCrossover(Crossover):
    def __init__(self):
        super().__init__(2, 1)

    def _do(self, problem, X, **kwargs):
        n_parents, n_matings, n_var = X.shape
        _X = np.full((self.n_offsprings, n_matings, problem.n_var), False)

        for k in range(n_matings):
            p1, p2 = X[0, k], X[1, k]

            both_are_true = np.logical_and(p1, p2)
            _X[0, k, both_are_true] = True

            n_remaining = problem.n_max - np.sum(both_are_true)

            I = np.where(np.logical_xor(p1, p2))[0]

            S = I[np.random.permutation(len(I))][:n_remaining]
            _X[0, k, S] = True

        return _X


class MyMutation(Mutation):
    def _do(self, problem, X, **kwargs):
        for i in range(X.shape[0]):
            X[i, :] = X[i, :]
            is_false = np.where(np.logical_not(X[i, :]))[0]
            is_true = np.where(X[i, :])[0]
            X[i, np.random.choice(is_false)] = True
            X[i, np.random.choice(is_true)] = False

        return X


# create the actual problem to be solved
np.random.seed(1)
L = np.array([[9, 13], [9, 14], [7, 15], [6, 12], [8, 16], [8, 14], [6, 12], [3, 18], [4, 14], [5, 14], [8, 11], [8, 18]])
n_max = 5
problem = SubsetProblem(L, n_max)

algorithm = NSGA2(
    pop_size=10,
    sampling=MySampling(),
    crossover=BinaryCrossover(),
    mutation=MyMutation(),
    eliminate_duplicates=True)

res = minimize(problem,
               algorithm,
               ('n_gen', 60),
               seed=1,
               verbose=True)

print(res.X)
print(res.F)

```

**Reply by blankjul** [reply]
is this what you are trying to do?

**Reply by usmanustc** [reply]
I have run it, and it works fine. Thanks for adjusting my code. The problem statement is:

Find a subset of 5 from this list where rank and mpg both are maximized. The way it is done however is, we are using the sum of all ranks and mpg in each subset to decide this. That is, it will pick 5 samples, and add their ranks and mpgs. Does it make sense if we talk about the original problem statement?

3 more points
1) When I output Sampling, it is only called once throughout the process. Is that correct?
2) How can I find the original subset (optimal one) used in the calculation of sums? 
print("Optimal Subset:", opt) -- this statement outputs [[0 1]
 [0 1]
 [0 1]
 [0 1]
 [0 1]]
3) Which changes are required to make this a maximization case.

**Reply by blankjul** [reply]
1. Yes!
2. `res.X` gives you all non-dominated solutions. It is bi-objective so there are multiple optima.
3. Just do `out["F"] = -1 * self.L[x].sum(axis=0).astype(float)` if both should be max.

**Reply by usmanustc** [reply]
I assume that this addresses the problem statement that I have shared above. Because doing sums is another way of saying that these 5 subsets present most optimum ranks and mpgs.

**Reply by blankjul** [reply]
I think it does. can you provide an exact mathematical formulation of your problem if you have doubts?

**Reply by usmanustc** [reply]
There is no exact mathematical formulation. The actual problem is somewhat similar.

I have a set of 10 classifiers. I have run them and calculated their results for an imbalanced data set (binary classification). I want to select 3 of them which present better performance on both classes. You will understand that some classifiers might present awful performances against a particular class. I have their Class0 Recall (Recall0), and Class1 Recall (Recall1). Simple permutation gives me 120 such subsets. I want to select a subset that performs better for both recalls.

**Reply by usmanustc** [reply]
What is your suggestion?

**Reply by blankjul** [reply]
why do optimization in the first place? Just get the 120 possible solutions and do non-dominated sorting. From the non-dominated set, you can then select the ones with different trade-offs.

**Reply by usmanustc** [reply]
So as  per theory "sort the solutions according to the Pareto dominance principle" ? Which parts of the above example need to be removed in that case or modified?

**Reply by blankjul** [reply]
You don't call an optimizer at all but use `NonDominatedSorting`. It is basically a totally different code compared to the above (but will give you the guaranteed optimum). If you are not familiar with the domination principle and how finding a non-dominated set of solutions works, I highly recommend: https://www.wiley.com/en-us/Multi+Objective+Optimization+using+Evolutionary+Algorithms-p-9780471873396

**Reply by usmanustc** [reply]
I get you. I understand that I need to find non-dominated solutions and then select the best ones. 

Could I use Pymoo for this purpose? https://pymoo.org/algorithms/moo/nsga2.html

**Reply by usmanustc** [reply]
thanks for the book suggestion. I will check it out.

**Reply by blankjul** [reply]
You can directly use `NonDominatedSorting` (https://github.com/anyoptimization/pymoo/blob/main/pymoo/util/nds/non_dominated_sorting.py). 
`NonDominatedSorting().do(F, only_non_dominated_front=True) `
